scripts/py_matter_yamltests/matter_yamltests/yaml_loader.py

```diff
@@ -0,0 +1,209 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
+#
+#    Licensed under the Apache License, Version 2.0 (the 'License');
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an 'AS IS' BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from typing import Union
+
+from .fixes import add_yaml_support_for_scientific_notation_without_dot
+
+try:
+    from yaml import CSafeLoader as SafeLoader
+except:
+    from yaml import SafeLoader
+
+import yaml
+
+
+class YamlLoader:
+    """This class loads a file from the disk and validate that the content is a well formed yaml test."""
+
+    def load(self, yaml_file: str) -> tuple[str, Union[list, str], dict, list]:
+        name = ''
+        pics = None
+        config = {}
+        tests = []
+
+        if yaml_file:
+            with open(yaml_file) as f:
+                loader = SafeLoader
+                add_yaml_support_for_scientific_notation_without_dot(loader)
+                content = yaml.load(f, Loader=loader)
+
+                self.__check_content(content)
+
+                name = content.get('name', '')
+                pics = content.get('PICS')
+                config = content.get('config', {})
+                tests = content.get('tests', [])
+
+        return (name, pics, config, tests)
+
+    def __check_content(self, content):
+        schema = {
+            'name': str,
+            'PICS': (str, list),
+            'config': dict,
+            'tests': list,
+        }
+
+        self.__check(content, schema)
+
+        tests = content.get('tests', [])
+        for test in tests:
+            try:
+                self.__check_test_step(test)
+            except KeyError as e:
+                raise KeyError(f'An exception occured in step {test}.')
+            except TypeError as e:
+                raise TypeError(f'An exception occured in step {test}.')
```

Comment:
Should we include the exception on the string somehow so we know what failed?

Reply:
So...
I was unhappy with the current exception mechanism, as stack traces are not that useful when it comes to map it to the issue in the YAML file. So Instead of adding more informations to the stack trace I reworked all of that a little bit.

Now with the right handling (which I hope to land asap) it looks like this:
<img width="727" alt="image" src="https://user-images.githubusercontent.com/65897387/218432858-a902e63f-46d4-4519-a61c-c5de2e45ece5.png">

I digged a little bit in the `pyyaml` source, anything more elaborate may require to build a custom scanner (e.g if we want the **exact** original line where the error comes from. If we have the time for it it would be nice, but in the meantime, and imho, this is already a much better error reporting than showing the stack...